src/extension/wps/src/main/java/au/org/emii/gogoduck/download/ParallelDownloadManager.java

```diff
@@ -0,0 +1,171 @@
+            throw new RuntimeException("Unexpected system error retrieving download", e);
+        }
+    }
+
+    @Override
+    public void remove() {
+        try {
+            logger.debug(String.format("Deleting download from %s", previous.getURL()));
+            Files.deleteIfExists(previous.getPath());
+            localStorageAllocated -= previous.getSize();
+            downloadUpToStorageLimit();
+        } catch (IOException e) {
+            throw new RuntimeException(
+                String.format("Unexpected system error: unable to delete downloaded file %s", previous.getPath()), e);
+        }
+    }
+
+    private void downloadUpToStorageLimit() {
+        DownloadRequest next = unactionedQueue.peek();
+
+        while (next != null && (localStorageAllocated == 0 || wontExceedLocalStorageLimit(next))) {
+            final DownloadRequest request = unactionedQueue.pop();
+            logger.debug(String.format("Adding %s to download queue", request.getUrl()));
+            inProgressQueue.add(pool.submit(new DownloadCallable(request)));
+            localStorageAllocated += request.getSize();
+            next = unactionedQueue.peek();
+        }
+    }
+
+    private boolean wontExceedLocalStorageLimit(DownloadRequest next) {
+        return localStorageAllocated + next.getSize() < config.getLocalStorageLimit();
+    }
+
+    private class DownloadCallable implements Callable<Download> {
+        private final DownloadRequest request;
+
+        public DownloadCallable(DownloadRequest request) {
+            this.request = request;
+        }
+
+        @Override
+        public Download call() {
+            logger.debug(String.format("Downloading %s", request.getUrl()));
+            int attempt = 1;
+            Download result = download();
+            while (result instanceof DownloadError && ++attempt <= config.getDownloadAttempts()) {
+                logger.debug(
+                    String.format("Downloading %s %s%s attempt", request.getUrl(), attempt, Ordinal.suffix(attempt)));
+                result = download();
+            }
+            if (result instanceof DownloadError) {
+                logger.warn(String.format("Failed to download %s after %d tries", request.getUrl(), config.getDownloadAttempts()));
+            } else {
+                logger.debug(String.format("Downloaded %s", request.getUrl()));
+            }
+            return result;
+        }
+
+        private Download download() {
+            Path path = config.getDownloadDirectory().resolve(UUID.randomUUID().toString());
+
+            try {
+                downloader.download(request.getUrl(), path);
+                return new Download(request.getUrl(), path, request.getSize());
+            } catch (IOException e) {
+                logger.warn(String.format("Could not download %s", request.getUrl()), e);
+                return new DownloadError(request.getUrl(), path, request.getSize(), e);
```

Comment:
It attempts to do the download up to the configured amount of times and then returns an error.  Its up to the client code what it does when an error happens - i.e gogoduck